### PR TITLE
[SPARK-8826] [SQL] Fix ClassCastException in GeneratedAggregate

### PR DESCRIPTION
**Pull request by navis**
When codegen is disabled, ClassCastException is thrown in some cases.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala [hunk 1]
```diff
@@ -22,9 +22,15 @@ package org.apache.spark.sql.catalyst.expressions
...
-    this(expressions.map(BindReferences.bindReference(_, inputSchema)))
+class InterpretedProjection(expressions: Seq[Expression], mutableRow: Boolean = false)
+  extends Projection {
+
+  def this(
+      expressions: Seq[Expression],
+      inputSchema: Seq[Attribute],
+      mutableRow: Boolean = false) = {
+    this(expressions.map(BindReferences.bindReference(_, inputSchema)), mutableRow)
+  }
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala]: Why this change? If you want to use mutable row in projection, you can use `InterpretedMutableProjection`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/GeneratedAggregate.scala [hunk 2]
```diff
@@ -241,7 +241,7 @@ case class GeneratedAggregate(
     child.execute().mapPartitions { iter =>
       // Builds a new custom class for holding the results of aggregation for a group.
       val initialValues = computeFunctions.flatMap(_.initialValues)
-      val newAggregationBuffer = newProjection(initialValues, child.output)
+      val newAggregationBuffer = newProjection(initialValues, child.output, mutableRow = true)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/GeneratedAggregate.scala]: We wrote `newAggregationBuffer(EmptyRow).asInstanceOf[MutableRow]` below, I think that is the root case. cc @davies

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala (hunk 1 above)
Comment (navis) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala]: Yes, but InterpretedProjection and InterpretedMutableProjection has different semantic on returning object. It would be same thing if call setTarget() before each apply() call but I wanted keep it simple without incurring another problem.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/AggregateSuite.scala [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import org.apache.spark.sql.SQLConf
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.test.TestSQLContext
+import org.apache.spark.sql.types.DataTypes._
+
+class AggregateSuite extends SparkPlanTest {
+
+  test("SPARK-8826 Fix ClassCastException in GeneratedAggregate") {
+
+    // when codegen = false, CCE is thrown if group-by expression is empty or unsafe is disabled
```
Comment (JoshRosen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/AggregateSuite.scala]: Doesn't the use of `GeneratedAggregate` imply that codegen must be enabled?

Comment (JoshRosen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/AggregateSuite.scala]: Basically, I'm wondering if this is a bug.  If this problem only occurs when mis-using GeneratedAggregate, then I think we should resolve this by adding an assertion / precondition check to GeneratedAggregate to ensure that codgen is enabled.

Comment (JoshRosen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/AggregateSuite.scala]: @navis, can you comment on this?

**Comment by JoshRosen** [comment]
I've closed the [JIRA issue](https://issues.apache.org/jira/browse/SPARK-8826) as "invalid" since GeneratedAggregate will never be used when codegen is disabled.  As a result, do you mind closing this issue?  Thanks!

**Comment by JoshRosen** [comment]
Bump; can you please close this pull request?  Thanks!
